modules/reactor/src/meshgenerators/AssemblyMeshGenerator.C

```diff
@@ -0,0 +1,503 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "AssemblyMeshGenerator.h"
+
+#include "MooseApp.h"
+#include "Factory.h"
+#include "libmesh/elem.h"
+#include "MooseMeshUtils.h"
+
+registerMooseObject("ReactorApp", AssemblyMeshGenerator);
+
+InputParameters
+AssemblyMeshGenerator::validParams()
+{
+  auto params = MeshGenerator::validParams();
+
+  params.addRequiredParam<std::vector<MeshGeneratorName>>(
+      "inputs", "The PinMeshGenerators that form the components of the assembly.");
+
+  params.addRequiredParam<subdomain_id_type>("assembly_type",
+                                             "The integer ID for this assembly type definition");
+
+  params.addRequiredParam<std::vector<std::vector<unsigned int>>>(
+      "pattern",
+      "A double-indexed array starting with the upper-left corner where the index"
+      "represents the layout of input pins in the assembly lattice.");
+
+  params.addRangeCheckedParam<std::vector<Real>>(
+      "duct_halfpitch",
+      "duct_halfpitch>0.0",
+      "Distance(s) from center to duct(s) inner boundaries.");
+
+  params.addRangeCheckedParam<unsigned int>("background_intervals",
+                                            "background_intervals>0",
+                                            "Radial intervals in the assembly peripheral region.");
+
+  params.addRangeCheckedParam<std::vector<unsigned int>>(
+      "duct_intervals", "duct_intervals>0", "Number of meshing intervals in each enclosing duct.");
+
+  params.addParam<std::vector<subdomain_id_type>>(
+      "background_region_id",
+      "The region id for the background area between the pins and the ducts");
+
+  params.addParam<std::vector<std::vector<subdomain_id_type>>>(
+      "duct_region_ids", "The region id for the ducts from innermost to outermost.");
+
+  params.addParam<bool>("extrude",
+                        false,
+                        "Determines if this is the final step in the geometry construction"
+                        " and extrudes the 2D geometry to 3D. If this is true then this mesh "
+                        "cannot be used in further mesh building");
```

Comment:
```suggestion
                        "cannot be used in a CoreMeshGenerator");
```
you could still stitch it to another assembly

Reply:
or see other comment for another formulation